### DISCUSSION TITLE
useIsError hook

### DISCUSSION THREAD

**Discussion by sarmong** [TanStack/query#2524] [Ideas]
I think it would be a good idea to implement another method in `QueryClient` and a new hook.
They will act in the same way as `useIsFetching` and `queryClient.isFetching` -- returning the number of currently failed queries.

Why is this useful?

To have one point for all the error handling and showing one error screen for the whole application

**Comment by sarmong** [comment]
I opened a PR adding this - #2525

**Comment by TkDodo** [comment]
> To have one point for all the error handling and showing one error screen for the whole application

If that's your goal, I think there are already two solutions to this problem:

1) `useErrorBoundary`

you don't need suspense to use error boundaries, you can use it to wrap a component tree in an error boundary and react-query will throw errors during the next render cycle so that you can handle all errors, including fetching errors, for that sub-tree in the error boundary.

2) `onError` callback on the query cache

The global callbacks are documented here: https://react-query.tanstack.com/reference/QueryCache. They are only called once per query, not per observer, and cannot be overwritten (as opposed to `defaultOptions`, which can be overwritten). You could use this to show error notifications, or even error screens on the level of the whole QueryClient.

Would either one of these solve your problem? I don't see much use for another hook that just gives you the number of erroneous queries, that doesn't seem very informative, because I won't even have access to the actual error message ....

**Reply by sarmong** [reply]
Thank you for the answer, @TkDodo .

What `useErrorBoundary` are you referring to? React's Error Boundary only resolves render errors, so it won't work.

`onError` could be a solution, but `QueryCache` is defined outside of react component. But I need to set state of my application by using `useContext`.

Also, I have the similar problem with Loading indicator. I need to have a Loading Screen show up whenever something is loading (but not fetching in the background!). Is there a neat solution to this without adding check for each query I have throughout the application?

**Reply by TkDodo** [reply]
> What useErrorBoundary are you referring to? React's Error Boundary only resolves render errors, so it won't work.

React Query has an option `useErrorBoundary` (the name is kinda bad because it sounds like a hook 😅 ) that you can pass to your query, or also set as default on the QueryClient:

```
useQuery(key, fn, { useErrorBoundary: true })
```

with this on, react-query will catch any fetching errors and re-throw them in the next render cycle, so that a react error boundary can pick it up. This is also meant for suspense, which has this on per default, but can just as easily be used without it. It means that you can use normal ErrorBoundaries to catch fetch errors, which is kinda cool.

> onError could be a solution, but QueryCache is defined outside of react component. But I need to set state of my application by using useContext.

you can also define your QueryClient inside your app, as long as you make sure that it's stable, e.g. by using state:

```
const App = () => {
  const [queryClient] = React.useState(() => new QueryClient())
  
  return <QueryClientProvider ...>
}
```

now `queryClient` will also only be created once, but you can use other hooks and use the result in the `onError` callback.

> Also, I have the similar problem with Loading indicator. I need to have a Loading Screen show up whenever something is loading (but not fetching in the background!). Is there a neat solution to this without adding check for each query I have throughout the application?

I think this is what suspense is for. Check out the example: https://react-query.tanstack.com/examples/suspense

**Reply by sarmong** [reply]
I will try the error solutions, thanks.

Regarding Suspense - I think it won't work for my situation. 
We have `App` component. Inside it is a nested structure of components that have their own rules for rendering. And these components have their own `useQuery`s. So, queries are all over the application. However, I need to show Error screen or Loading screen on the `App` level. 

Currently there is an `AppStateContext.Provider` that wraps the application. And for each component with query there is a check for `isError` and `isLoading` and if true - they send update app state with `useContext` to show error/loading screen.

So, whenever new query appears, new check has to be added. This would be prone to bugs and not good architecture.

With regard to `React.Suspense` I would need to check each query, but it is not possible and would make a spaghetti code.

**Reply by TkDodo** [reply]
> So, queries are all over the application. However, I need to show Error screen or Loading screen on the App level.

As far as I know this is _exactly_ where suspense shines. You wrap your whole app in a `<Suspense fallback="loading...">`and whenever a query wants to load data, the suspense fallback will be shown. Not sure what checks would need to be added there?

**Reply by TkDodo** [reply]
@michalenr  are you fine with closing the PR if you’ve found another way to fix your use case?

**Reply by sarmong** [reply]
Sorry for late response. `useErrorBoundary` is what I needed, thank you. I will close the PR.

Regarding Suspence - I didn't try it yet. But won't it be using fallback also when the query is fetching in the background?

**Reply by sarmong** [reply]
@TkDodo , so Suspense is an experimental api and I wouldn't want to use it for production.

Would it make sense to make a hook `useIsLoading` that would work in a similar way as `useIsFetching` ?

**Reply by sarmong** [reply]
Or at least have `isLoading` key in `queryClient.getQueryState`

**Reply by TkDodo** [reply]
> Or at least have isLoading key in queryClient.getQueryState

that function returns the `state` the query is in, which is:

```
'idle' | 'loading' | 'error' | 'success'
```

the `isLoading` etc boolean are just derived from that state.

**Reply by ggregoire** [reply]
`useIsLoading` would be a great quality of life addition!

(I just found that thread searching `react-query global isLoading` in google 😅)
